src/main/java/com/mraof/minestuck/entity/AnimatedCreatureEntity.java

```diff
@@ -0,0 +1,165 @@
+import net.minecraft.network.datasync.DataSerializers;
+import net.minecraft.network.datasync.EntityDataManager;
+import net.minecraft.world.World;
+import software.bernie.geckolib3.core.IAnimatable;
+import software.bernie.geckolib3.core.controller.AnimationController;
+import software.bernie.geckolib3.core.manager.AnimationFactory;
+
+public abstract class AnimatedCreatureEntity extends CreatureEntity implements IAnimatable {
+    private final AnimationFactory factory = new AnimationFactory(this);
+    protected static final DataParameter<Integer> CURRENT_ACTION = EntityDataManager.defineId(AnimatedCreatureEntity.class, DataSerializers.INT);
+
+    private int heavyAttackTicks = 0;
+    private int recoveryTicks = 0;
+    private int attackDelay;
+    private int attackRecovery;
+    private int actionResetTicks;
+
+    protected boolean canMoveWhileAttacking = false;
+
+    protected AnimatedCreatureEntity(EntityType<? extends CreatureEntity> type, World world) {
+        super(type, world);
+    }
+
+    @Override
+    protected void registerGoals() {
+        super.registerGoals();
+        goalSelector.addGoal(3, new DelayedAttackGoal(this, 1F, false));
+    }
+
+    @Override
+    protected void defineSynchedData() {
+        super.defineSynchedData();
+        entityData.define(CURRENT_ACTION, Actions.NONE.ordinal());
+    }
+
+    @Override
+    public void tick() {
+        super.tick();
+        if (heavyAttackTicks > 0) {
+            heavyAttackTicks--;
+            if (heavyAttackTicks == 0) {
+                recoveryTicks = attackRecovery;
+                if (getTarget() != null && isInRange(getTarget())) { // TODO: AOE bounding box collision checks + aoe flag
+                    doHurtTarget(getTarget());
+                }
+            }
+        }
+        if (recoveryTicks > 0) {
+            recoveryTicks--;
+            if (recoveryTicks == 0) {
+                setAttacking(false);
+            }
+        }
+        if (getCurrentAction() != Actions.NONE && actionResetTicks > 0) {
+            actionResetTicks--;
+            if (actionResetTicks <= 0) {
+                setCurrentAction(Actions.NONE, 0);
+            }
+        }
+    }
+
+    private boolean isInRange(LivingEntity target) {
+        double reach = this.getBbWidth() * 2.0F * this.getBbWidth() * 2.0F + target.getBbWidth();
+        return this.distanceToSqr(target) <= reach;
+    }
+
+    private void startAttack() {
+        if (heavyAttackTicks <= 0 && recoveryTicks <= 0) {
+            heavyAttackTicks = attackDelay;
+            setAttacking(true);
+        }
+    }
+
+    protected AnimationController<AnimatedCreatureEntity> createAnimation(String name, double speed, AnimationController.IAnimationPredicate<AnimatedCreatureEntity> predicate) {
+        AnimationController<AnimatedCreatureEntity> controller = new AnimationController<>(this, name, 0, predicate);
+        controller.setAnimationSpeed(speed);
+        return controller;
+    }
```

Comment:
It is intelligent to use this method of controlling the speed of the animation, but then are the conditions in which the speed is not 1 suggest its not being used as intended upon creation of the animation? Let me know if I should reword that question

Reply:
can you reword please? its intended to be used with or without speed

Reply:
Reworded and resolved outside of PR